velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -321,10 +321,40 @@ bool NestedLoopJoinProbe::addToOutput() {
     }
 
     // Only re-calculate the filter if we have a new build vector.
-    if (buildRow_ == 0) {
+    if (buildRow_ == 0 && joinCondition_ != nullptr) {
       evaluateJoinFilter(currentBuild);
     }
 
+    //
+    // Implements a a Left Semi Project Join within NestedLoopJoinProbe.
+    // The getOutputLeftSemiJoinImpl() will ensure that exactly one row is
+    // produced for each probe row, along with a boolean "match" column
+    // which will indicate whether a matching build row exists on the
+    // build side.
+    //
+    // 1. At this point, the filter expressions are applied and we short
+    //    circuit the execution for a LeftSemiProject since we don't require
+    //    mismatch rows or build side projections. For each probe row, we simply
+    //    iterate through decoded filter results to determine if at least
+    //    one build side row satisfies the filter condition.
+    // 2. If match is found, the match column is marked as `true`, and
+    //    defaulted to false otherwise. Finally populates the output row with
+    //    the probe row data.
+    // 3. The function ensures that only one row is produced in the output,
+    //    indicating build side match. After processing the current probe row,
+    //    it skip the rest of the build rows.
+    //
+    // Returns a `RowVectorPtr` representing the output row. For left semi project
+    // this basically contains probe row data with the match column.
+    //
```

Comment:
Perhaps move this documentation to 'NestedLoopJoinProbe.h' for the getOutputLeftSemiJoinImpl function.